### DISCUSSION TITLE
ioc.fake must come before any `use()`?

### DISCUSSION THREAD

**Discussion by otang** [adonisjs/core#1750] [Help] [answered]
In many of my modules I've opted to use `use('App/Services/MyService')` instead of requiring it. However I have not registered these services with `ioc.bind`, but instead allowing auto-loading to take place. Typically the `use()` statements are at the top of my files.

In my tests, I'm trying to use `ioc.fake` to stub a method of one of my services, but I've found that if my `use()` statement comes at the top of the file, before the test runs, it is unable to successfully stub. The alternative is moving the `use()` statement into another method so that it's only evaluated at runtime and after the stub. It doesn't seem like a good solution to move all `use()` statements out of the global scope.

What am I doing wrong here?

Secondly, in the docs there are many examples of registering dependencies with `ioc.bind`, however it doesn't suggest where this should be done. Is there an opinionated place to register all the `ioc.binds()`?

**Comment by thetutlage** [comment, marked as answer]
Yup, you are spot on with this. The `fake` method is not really that helpful with v4 (we have fixed it in v5 though). Once has module has been imported (ie the `use` method has been used), the fake cannot undo its imported value.

With v5 we are using proxies to fix this behaviour and works like a charm, but back porting it lead to breaking changes.

Regarding, where to use `ioc.bind`? The answer is "It depends"

- If you are creating a package, then write it inside the provider file
- If you are in your application codebase, again you can create a provider in the same codebase and write it inside the provider file. However, you should avoid binding stuff to the container, when you can actually import it directly by passing the path of the file (as you are doing it right now `App/Services/MyService`.
